feat(curriculum): add express middleware lessons (#67194)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/blocks/lecture-express-middleware/69f42d414b7626e21a922794.md

---
id: 69f42d414b7626e21a922794
title: What is Error-handling Middleware?
challengeType: 19
dashedName: what-is-error-handling-middleware
---

# --description--

Error-handling middleware in Express is a special type of middleware used to handle errors that occur during the request-response cycle. It's essential for catching unexpected errors, logging them, and providing appropriate responses to the client.

> In Express, error-handling middleware has four parameters: `err`, `req`, `res`, and `next`. This middleware is invoked when an error is passed to the `next()` function.

Reviewer: Are we intentionally duplicating the instructional content?

https://github.com/freeCodeCamp/freeCodeCamp/pull/67192/changes#diff-732c1f01f50423b0ba5913f506966d6377847fb365f94db0eb3a19c206f37deaR18

Author: I'm sure it's not intentional. The same person wrote the lectures, so it's possible the same concept came up naturally in both lessons. I will push a fix to this one cuz it comes after the other one.

Author: I'm also going to remove the error-handling middleware sample code in the other one, and leave a note that learners will learn more about error-handling middleware here, so this one looks more unique.

})
```

* Route Handler: The route handler intentionally throws an error to simulate a problem.
* Error-handling Middleware: This middleware is registered with `app.use()`. It catches errors passed via `next(err)` and sends a 500 HTTP response to the client with a generic error message.
* `err`: This is the error object that gets passed to the middleware.

Error-handling middleware ensures that when something goes wrong in your application, users still get a proper response. It's important to always include this middleware at the end of your middleware stack to catch any errors.

# --questions--

## --text--

What does error-handling middleware do?

## --answers--

It processes requests and sends responses.

### --feedback--

Focus on what happens when something goes wrong during a request.

---

It handles errors that occur during the request-response cycle.

---

It serves static files.

### --feedback--

Focus on what happens when something goes wrong during a request.

---

It handles routing logic.

### --feedback--

Focus on what happens when something goes wrong during a request.

## --video-solution--

2

## --text--

> Which of the following is required in error-handling middleware?

Reviewer: ```suggestion
Which of the following parameters are required in error-handling middleware?
```

## --answers--

`err`, `req`, `res`, `next`.

---

`req` and `res`.

### --feedback--

Think about the parameters Express uses to recognize error handlers.

---

`err` and `next`.

### --feedback--

Think about the parameters Express uses to recognize error handlers.

---

`err` and `req`.

### --feedback--

Think about the parameters Express uses to recognize error handlers.

## --video-solution--

1

## --text--

Where should error-handling middleware be placed in the middleware stack?

## --answers--

At the top, before other middleware.

### --feedback--

> Think about where error handlers are placed in a code that might fail.

Reviewer: ```suggestion
Think about where error handlers are placed in code that might fail.
```

---

In the middle of the stack.

### --feedback--

> Think about where error handlers are placed in a code that might fail.

Reviewer: ```suggestion
Think about where error handlers are placed in code that might fail.
```

---

At the bottom, after other middleware.

---

There is no specific place for it.

### --feedback--

> Think about where error handlers are placed in a code that might fail.

Reviewer: ```suggestion
Think about where error handlers are placed in code that might fail.
```
